### DISCUSSION TITLE
Update UserId - Is it advisable?

### DISCUSSION THREAD

**Discussion by orchardc** [OrchardCMS/OrchardCore#8333] [Q&A]
I know the UserId is supposed to be 'immutable', but we have a use case to update it on daily basis per user (for about 10,000 users)

Is it advisable? 

Thanks!

**Comment by orchardc** [comment]
@deanmarcussen any idea?

**Reply by Skrypt** [reply]
As long as you also update every related DB records at the same time.

**Reply by orchardc** [reply]
Many thanks @Skrypt , we use the code below to update the UserId, the code updates two tables in one go: UserIndex and Document.
Is there anything else must do for every related DB records ?

`myUser.UserId = newId;
await _userManager.UpdateAsync(myUser);`

**Comment by deanmarcussen** [comment]
Can’t recommend it. Defeats the point of the userid. You would have to force break all the cookies as well as they would no longer match.

**Reply by orchardc** [reply]
Thanks @deanmarcussen , In our case, it doesn't require cookies. we simply need a solution to cover the following features:
1, set a unique id on the user's record, which will then expire in a day.
2, there must be an efficient way to retrieve the user's object by our unique id; so in this case, we just use: `_userManager.FindByIdAsync(MyUniqueId) as User;`

Not sure if there is a better solution?

**Reply by deanmarcussen** [reply]
The UserId is an immutable record of the UserId.

I cannot recomend changing it every day. 

Suggest:
Make your own Guid, put it on the entity properties. Create an index. query the index as required.
```
var idWrapper = new IdWrapper { CustomId = Guid. NewGuid() }
user.Put(idWrapper);
... save etc...
```

Note: if this is going anywhere near a browser, don't use a guid. use a time limited data protector.
https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/consumer-apis/limited-lifetime-payloads?view=aspnetcore-5.0

Better yet: don't save anything on the user at all, and just encode someting in a time limited dataprotector

**Reply by orchardc** [reply]
Thanks @dalenewman , but how to query by this GUID and return the user object? I saw your example yesterday in the CustomSetting, but I'm not clear about how to query it, is it possible to use the example you provided here to show how to query it?

**Reply by deanmarcussen** [reply]
https://github.com/sebastienros/yessql/wiki/Tutorial#querying-documents

**Reply by orchardc** [reply]
Thanks @deanmarcussen , exactly what I'm looking for! :)
